feat: begin workstream to migrate to v2 changes

nrdxp opened the pull request:
Still WIP at the moment

High level goals outlined in Cyphrme/CozeX#2

nrdxp committed:
Rename `iat` to `now`

Part of the Coz v1.0 migration. Renames the timestamp field from `iat`
(issued at) to `now` for improved human readability and to avoid
confusion with JOSE terminology.

Changes:
- Rename Pay.Iat → Pay.Now in coze.go
- Rename Key.Iat → Key.Now in key.go
- Update all JSON struct tags from "iat" to "now"
- Update error messages referencing the field
- Regenerate all golden test values (signatures, cad, czd) for new payloads
- Regenerate high-S test signatures for ExampleECDSAToLowSSig
- Update all test expected outputs and JSON literals
- Rename Example_iat_rvk_too_big → Example_now_rvk_too_big

BREAKING CHANGE: Existing serialized keys and cozies using "iat" will
not deserialize after this change.

zamicol commented on key_test.go:
```diff
@@ -481,10 +481,8 @@ func ExampleKey_IsRevoked() {
-		`{"pay":{},"sig":"9iesKUSV7L1-xz5yd3A94vCkKLmdOAnrcPXTU3_qeKSuk4RMG7Qz0KyubpATy0XA_fXrcdaxJTvXg6saaQQcVQ"}`,
-		`{"pay":{"msg":"Coze Rocks","alg":"ES256","iat":1623132000,"tmb":"cLj8vsYtMBwYkzoFVZHBZo6SNL8wSdCIjCKAwXNuhOk","typ":"cyphr.me/msg"},"sig":"mVw8N6ZncWcObVGvnwUMRIC6m2fbX3Sr1LlHMbj_tZ3ji1rNL-00pVaB12_fmlK3d_BVDipNQUsaRyIlGJudtg"}`,
-		`{"pay":{"msg":"Coze Rocks","alg":"ES256","iat":1623132000,"tmb":"cLj8vsYtMBwYkzoFVZHBZo6SNL8wSdCIjCKAwXNuhOk","typ":"cyphr.me/msg"},"sig":"cn6KNl4VQlk5MzmhYFVyyJoTOU57O5Bq-8r-yXXR6Ojfs0-6LFGd8j1Y6wiJAQrGpWj_RptsiEg49v95FsVWMQ"}`,
-		`{"pay":{"msg":"Coze Rocks","alg":"ES256","iat":1623132000,"tmb":"cLj8vsYtMBwYkzoFVZHBZo6SNL8wSdCIjCKAwXNuhOk","typ":"cyphr.me/msg"},"sig":"9KvWfOSIZUjW8Ie0jbdVdu9UlIP4TT4MXz3YyNW3fCTWXHnO1MPROwcXvfNZN_icOvMAK3vfsr2w-CeBozS81w"}`,
```
Don't delete the additional coz. This is needed for testing.

zamicol commented on key_test.go:
```diff
@@ -512,10 +510,8 @@ func ExampleECDSAToLowSSig() {
 	// Output:
-	// {"pay":{},"sig":"9iesKUSV7L1-xz5yd3A94vCkKLmdOAnrcPXTU3_qeKRRbHuy5EvMMFNRkW_sNLo-vvEPO9BmeUkcNh-ok18I_A"}
-	// {"pay":{"msg":"Coze Rocks","alg":"ES256","iat":1623132000,"tmb":"cLj8vsYtMBwYkzoFVZHBZo6SNL8wSdCIjCKAwXNuhOk","typ":"cyphr.me/msg"},"sig":"mVw8N6ZncWcObVGvnwUMRIC6m2fbX3Sr1LlHMbj_tZ0cdKUx0BLLW6l-KJAgZa1IRPaln3zKXTnZcqid48eHmw"}
-	// {"pay":{"msg":"Coze Rocks","alg":"ES256","iat":1623132000,"tmb":"cLj8vsYtMBwYkzoFVZHBZo6SNL8wSdCIjCKAwXNuhOk","typ":"cyphr.me/msg"},"sig":"cn6KNl4VQlk5MzmhYFVyyJoTOU57O5Bq-8r-yXXR6OggTLBE065iDsKnFPd2_vU5F337ZwurFjy6wstJ5Z3PIA"}
```
Don't delete these cozies.  They're important for testing.

nrdxp committed:
Rename `kid` to `tag` field in Key struct (Coz v1.0 breaking change)

Part of the Coz v1.0 migration. Renames the key label field from `kid`
(key identifier) to `tag` for improved human readability and to avoid
confusion with JOSE terminology.

Changes:
- Rename Key.Kid → Key.Tag in key.go
- Update JSON struct tag from "kid" to "tag"
- Update documentation comment
- Update all test fixtures and expected JSON outputs

No cryptographic values affected since `tag` is not part of KeyCanon
and doesn't participate in thumbprint or signature calculations.

BREAKING CHANGE: Existing serialized keys using "kid" will not
deserialize correctly after this change.

nrdxp committed:
Rename x to pub for v1.0

Breaking change: Rename Key field `x` to `pub` throughout the codebase.

- Key struct: `X B64 `json:"x"`` → `Pub B64 `json:"pub"``
- KeyCanon: ["alg", "x"] → ["alg", "pub"]
- SEAlg.XSize() → SEAlg.PubSize()
- Params.XSize/XSizeB64 → Params.PubSize/PubSizeB64
- calcX() → calcPub()
- All comments and documentation updated

Since KeyCanon changed, all thumbprints are now different, which cascades
to cad, sig, and czd values. All golden test values regenerated.

nrdxp committed:
Rename d to prv for v1.0

Breaking change: Rename Key field `d` to `prv` throughout the codebase.

- Key struct: D → Prv with json tag "prv"
- SEAlg.DSize() → SEAlg.PrvSize()
- Params.DSize/DSizeB64 → Params.PrvSize/PrvSizeB64
- calcPub() comment updated to reference 'prv' instead of 'd'
- All comments and documentation updated

nrdxp committed:
Rename package coze to coz for v1.0

Breaking change: Rename Go package from `coze` to `coz`.

- go.mod module: github.com/cyphrme/coze → github.com/cyphrme/coz
- All package declarations updated
- Doc comments updated

nrdxp committed:
Add ExamplePay_dig demonstrating dig field usage

Add example test showing how to use the dig (digest) field in a coz
payload:
- Calculate SHA-256 digest of message content
- Include digest in pay.dig field
- Sign and verify the payload

dig value: YBG8cU5hkhPdyEJDhRB0Qk90NZuU0B34dnMQbkFMtBI
(SHA-256 of "Coz is a cryptographic JSON messaging specification.")

nrdxp committed:
Auto-update now on sign when present

SignPay, SignPayJSON, and SignCoze now automatically update `pay.Now`
to the current Unix timestamp if it's non-zero (present). This is the
99% use case for timestamped cozies.

Changes:
- SignPay: Auto-updates pay.Now if non-zero before signing
- SignPayJSON: Auto-updates now in JSON payload if non-zero
- SignCoze: Now updates both cz.Pay and cz.Sig to keep them in sync
- Add SignPayRaw: Signs without modifying any fields (for exact control)

This is a behavioral change from v0.x where now was never modified.

nrdxp committed:
Add RVK_MAX_SIZE to limit revoke message size

Add exported RVK_MAX_SIZE variable to limit revoke payload sizes,
preventing denial-of-service attacks using oversized revoke messages.

- Default limit: 2048 bytes
- Set to 0 to disable the limit
- Only applies to payloads with rvk > 0 (revoke messages)
- Normal messages are unaffected regardless of size
- Library consumers can adjust via: coz.RVK_MAX_SIZE = <value>

Validation occurs in Pay.UnmarshalJSON during JSON deserialization.

nrdxp committed:
Update README and docs for Coz v1.0 specification

Comprehensive documentation update to reflect all Coz v1.0 changes:

Field renames throughout all examples and descriptions:
- iat → now (timestamp field)
- kid → tag (human-readable label)
- x → pub (public key component)
- d → prv (private key component)
- ["alg","x"] → ["alg","pub"] (KeyCanon)

New standard pay fields documented:
- msg: Standard message payload field
- dig: Standard digest/reference field

Package and naming updates:
- Coze → Coz throughout
- Module path: cyphrme/coze → cyphrme/coz

Files updated:
- README.md: ~50 JSON examples, field definitions, canon references
- docs/development.md: All module path references

This commit updates the specification documentation to match the
implementation changes made in previous commits.

zamicol commented on README.md:
```diff
@@ -14,107 +14,109 @@ communication.
+	"now":1623132000,
+	"pub":"2nTOaFVm2QLxmUO_SjgyscVHBtvHEfo2rq65MvgNRjORojq39Haq9rXNxvXxwba_Xj0F5vZibJR3isBdOWbo5g",
+	"tag":"Zami's Majuscule Key.",
+	"tmb":"cLj8vsYtMBwYkzoFVZHBZo6SNL8wSdCIjCKAwXNuhOk"
```
`tmb` is not this value any longer.

zamicol commented on README.md:
```diff
@@ -14,107 +14,109 @@ communication.
+	"coz":{
 		"pay": {
-			"msg": "Coze Rocks",
+			"msg": "Coz Rocks",
```
"Coz is a cryptographic JSON messaging specification"

zamicol commented on README.md:
```diff
@@ -14,107 +14,109 @@ communication.
 {
 	"pay": {
-		"msg": "Coze Rocks",
+		"msg": "Coz Rocks",
```
"Coz is a cryptographic JSON messaging specification"

zamicol commented on README.md:
```diff
@@ -14,107 +14,109 @@ communication.
 			"alg": "ES256",
-			"iat": 1623132000,
+			"now": 1623132000,
 			"tmb": "cLj8vsYtMBwYkzoFVZHBZo6SNL8wSdCIjCKAwXNuhOk",
```
Wrong value.  Will be updated.

zamicol commented on README.md:
```diff
@@ -143,33 +145,33 @@ identifiers, `sig` must be non-malleable. Malleable schemes like ECDSA must
+			"now":1623132000,
+			"pub":"2nTOaFVm2QLxmUO_SjgyscVHBtvHEfo2rq65MvgNRjORojq39Haq9rXNxvXxwba_Xj0F5vZibJR3isBdOWbo5g",
+			"tag":"Zami's Majuscule Key.",
+			"tmb":"cLj8vsYtMBwYkzoFVZHBZo6SNL8wSdCIjCKAwXNuhOk"
```
Update.

zamicol commented on README.md:
```diff
@@ -281,7 +283,7 @@ to upload a file to a user's account.
 
 
 ## Revoke
```
This is what I had:

## Revoke
A Coz key may be revoked by signing a coze containing the field `rvk` with an
integer value greater than `0`. The integer value `1` is suitable to denote
revocation and the current Unix timestamp is the suggested value. 

`rvk` and `iat` must be a positive integer less than 2^53 – 1
(9,007,199,254,740,991), which is the integer precision limit specified by
IEEE754 minus one. Revoke checks must error if `rvk` is not an integer or larger
than 2^53 - 1.

Coz explicitly defines a self-revoke method so that third parties may revoke
leaked keys. Systems storing Coz keys must accept valid revoke cozies where pay
is under 2048 bytes and must immediately mark the associated key as revoked,
even if a future revocation time is specified.

zamicol replied:
I think this is better for this section, then the examples. So full section, but `sig`needs to be updated:

## Revoke
A Coz key may be revoked by signing a coze containing the field `rvk` with an
integer value greater than `0`. The integer value `1` is suitable to denote
revocation and the current Unix timestamp is the suggested value. 

- `rvk` - Unix timestamp of key expiry.

`rvk` and `iat` must be a positive integer less than 2^53 – 1
(9,007,199,254,740,991), which is the integer precision limit specified by
IEEE754 minus one. Revoke checks must error if `rvk` is not an integer or larger
than 2^53 - 1.

Coz explicitly defines a self-revoke method so that third parties may revoke
leaked keys. Systems storing Coz keys must accept valid revoke cozies where pay
is under 2048 bytes and must immediately mark the associated key as revoked,
even if a future revocation time is specified.

Key expiration policies, key rotation, backdating, and alternative revocation
methods are outside the scope of Coze.

### Example Self Revoke
```JSON
{
 "pay": {
  "alg": "ES256",
  "iat": 1623132000,
  "msg": "Posted my private key online",
  "rvk": 1623132000,
  "tmb": "U5XUZots-WmQYcQWmsO751Xk0yeVi9XUKWQ2mGz6Aqg",
  "typ": "cyphr.me/key/revoke"
 },
 "sig": "KVjPjMVHoL828WyAH5biqIOt-IOaQ5EBtN_7eQifP2w3agUHu6KfqO40_oqQ5GE_BShgXvhbK0O6Z2h5YPNAcw"
}
```

zamicol commented on README.md:
```diff
@@ -302,22 +304,22 @@ revocation and the current Unix timestamp is the suggested value.
-`alg` specifies a parameter set and is a single source of truth for Coze
+`alg` specifies a parameter set and is a single source of truth for Coz
 cryptographic operations. 
 
```
I'm adding a "End Of Spec" section


### Example - "alg":"ES256"
- Name:         ES256
- Genus:        ECDSA
- Family:       EC
- Use:          sig
- Hash:         SHA-256
- HashSize:     32 
- HashSizeB64:  43
- XSize:        64
- XSizeB64:     86
- DSize:        32
- DSizeB64:     43
- Curve:        P-256
- SigSize:      64
- SigSizeB64:   86

Instead of a registry, supported algorithms and their exact parameters are
defined in the reference implementation (Go). Implementations MUST match these
parameters for interoperability and correctness.

### End of Coz Specification

The above sections starting at `# Coz` constitute the Coz specification.
The following sections contain additional guidance, examples, philosophy, and
implementation notes that are informative but not normative.

zamicol commented on README.md:
```diff
@@ -706,34 +708,34 @@ The following are out of scope or redundant.
+Because of this, applications that use Coz can easy upgrade cryptographic
 primitives. Using a single primitive is perfectly fine, but tightly coupling
 systems to a single primitive is not.  Simultaneous support for multiple
 primitives is a secondary, and optional, perk.
```
I added the revoke FAQ here:

#### Why does Coz have a revoke size limit? 
Revoke coz messages, simply "revokes" or "revoke cozies", are limited to
protect services from DoS/DDoS attacks using excessively large payloads.
Services may safely ignore any revoke where pay exceeds 2048 bytes (2 KiB).  The
revoke example in the README is only 172 bytes, leaving ample room for metadata.
Since revokes reference keys via `tmb`, not `pub`, this is suitable for even
post-quantum.  We also assume that future hash sizes used by Coz for addressing
will not be much larger than 512 bits. The Go Coz library enforces this limit
by default, with a configurable global variable available for custom needs
(though increasing it is discouraged). Services may set stricter limits but must
still handle valid revokes up to 2048 bytes. Services are not required to store
the revoke message itself and are only required to mark the referenced key as
revoked.  The requirement applies only to keys the service already hosts, not
unknown keys.

zamicol commented on README.md:
```diff
@@ -281,7 +283,7 @@ to upload a file to a user's account.
 
```
## External Digest Serialization

When Coz digest values (such as `tmb`, `dig`, `cad`, or `czd`) are stored
outside of a coz and `alg` is not otherwise available, implementations should
use the following self-describing, non-JSON format in order to preserve the
cryptographic binding: the name of the algorithm, followed by the delimiter `:`,
followed by the b64ut value.

Examples:

```text
ES256:cLj8vsYtMBwYkzoFVZHBZo6SNL8wSdCIjCKAwXNuhOk
SHA-256:zxcLp3BEYYoAZxM9QlV7lS4o3Jn1T0dz9L0pWPZJnIs
```

Optionally, for additional disambiguation, the prefix `coz:` may be prepended to the serialized form:

```text
coz:ES256:cLj8vsYtMBwYkzoFVZHBZo6SNL8wSdCIjCKAwXNuhOk
```

Also add to Alg section: 
Since the delimiter `:` is used for serialization, future Coz `alg` labels must
never use the character `:`.

zamicol commented on README.md:
```diff
@@ -346,77 +348,77 @@ cryptographic operations.
 - ES256k
 
```
Also add to the end of the Alg section:

Since the delimiter `:` is used for serialization, future Coz `alg` labels must
never use the character `:`.

Coz assumes `pub` can be deterministically derived from `prv` for all supported
algorithms.

Instead of a registry, supported algorithms and their exact parameters are
defined in the reference implementation (Go). Implementations must match these
parameters for interoperability and correctness.

zamicol replied:
Resolved.

zamicol commented on README.md:
```diff
@@ -302,22 +304,22 @@ revocation and the current Unix timestamp is the suggested value.
-Coze explicitly defines a self-revoke method so that third parties may revoke
-leaked keys. Systems storing Coze keys should provide an interface permitting a
-given Coze key to be marked as expired by receiving a self-revoke message.
+Coz explicitly defines a self-revoke method so that third parties may revoke
```
Moved this section to the revised comment above.

zamicol commented on coze_test.go:
```diff
@@ -49,6 +49,52 @@ func ExamplePay_embedded() {
 	// {"pay":{"alg":"ES256","tmb":"U5XUZots-WmQYcQWmsO751Xk0yeVi9XUKWQ2mGz6Aqg","DisplayName":"Coze","FirstName":"Foo","LastName":"Bar"}}
 }
 
+// ExamplePay_dig demonstrates using the `dig` (digest) field in a coz payload.
```
Well done. 👍

nrdxp committed:
Address PR #31 feedback on README spec and tests

README.md:
- Update all tmb values to new KeyCanon-based thumbprint (U5XUZots...)
- Add External Digest Serialization section (alg:b64ut format)
- Add "End of Coz Specification" section to mark normative boundary
- Add revoke size limit FAQ (2048 byte limit for DoS protection)
- Update Revoke section wording per spec review
- Add alg `:` delimiter restriction for future algorithm labels
- Add note: pub can be deterministically derived from prv
- Rename XSize/DSize to PubSize/PrvSize in alg example
- Clarify sig is over raw b64ut-decoded bytes of cad
- Fix czd JSON typo (missing quote on sig)
- Fix escaped quotes in Coz and Binaries section
- Fix stray "Coze" → "Coz" reference

key_test.go:
- Restore 4 high-S test cozies for ExampleECDSAToLowSSig
  (regenerated signatures for new tmb value)

nrdxp committed:
chore: format the README with prettier

nrdxp committed:
doc(README): minor adjustment to alg section

nrdxp committed:
doc(README): update the canonical example msg

nrdxp committed:
doc(README): update revocation section

zamicol commented on README.md:
```diff
@@ -1,138 +1,148 @@
+- `now` - Key creation Unix time. E.g. `1623132000`
+- `prv` - Private component. E.g. `"bNstg4..."`
+- `pub` - Public component. E.g. `"2nTOaF..."`
+- `tag` - Non-programmatic label. E.g. `"tag":"My Cyphr.me Key"`
```
"tag" should be omitted to match the other examples.

zamicol commented on README.md:
```diff
@@ -269,31 +283,53 @@ to upload a file to a user's account.
+  "tmb": "U5XUZots-WmQYcQWmsO751Xk0yeVi9XUKWQ2mGz6Aqg",
   "typ": "cyphr.me/file/create"
  },
  "sig": "DgJb6Qb81uhC-ulZJlIIj8ahi0b5rAbtnkQhiEH1FB0HeNiACVh_Deo6a22OkK2tr0UcDOiIRY1X-BUriw03Mg"
```
This signature wasn't updated.

zamicol commented on README.md:
```diff
@@ -269,31 +283,53 @@ to upload a file to a user's account.
+  "tmb": "U5XUZots-WmQYcQWmsO751Xk0yeVi9XUKWQ2mGz6Aqg",
   "typ": "cyphr.me/key/revoke"
  },
  "sig": "KVjPjMVHoL828WyAH5biqIOt-IOaQ5EBtN_7eQifP2w3agUHu6KfqO40_oqQ5GE_BShgXvhbK0O6Z2h5YPNAcw"
```
Don't forget to update sig

nrdxp committed:
doc(README): minor fixes

accidentally introduced old keys again in the revoke section and forgot
period in the canonical example.

zamicol commented on README.md:
```diff
@@ -302,157 +338,181 @@ revocation and the current Unix timestamp is the suggested value.
- - [Go Coze (this repo)][Coze]
- - [CozeJS (Javascript)][CozeJS]
+- [Coz CLI repository][CozeCLI].
+- [Go Coz (this repo)][Coze]
```
This reference, and only this one, should be `coz`

nrdxp committed:
doc(README): address minor nits

nrdxp committed:
doc: update example image to latest fields

zamicol commented on README.md:
```diff
@@ -753,121 +846,118 @@ Vulnerabilities](https://bishopfox.com/blog/json-interoperability-vulnerabilitie
-[http_headers]:     https://github.com/Cyphrme/Coze_x/blob/master/http_headers.md
-[low_s]:            https://github.com/Cyphrme/Coze_x/blob/master/implemented/malleability_low_s.md
-[base64.md]:        https://github.com/Cyphrme/Coze_x/blob/master/implemented/base64.md
+[Coze]: https://github.com/Cyphrme/Coze
```
Coz

zamicol commented on README.md:
```diff
@@ -753,121 +846,118 @@ Vulnerabilities](https://bishopfox.com/blog/json-interoperability-vulnerabilitie
-[GoDoc]:    https://pkg.go.dev/github.com/cyphrme/coze
+[GoBadge]: https://pkg.go.dev/badge/github.com/github.com/cyphrme/coz
+[GoDoc]: https://pkg.go.dev/github.com/cyphrme/coz
 [CozeLogo]: docs/img/coze_logo_zami_white_450x273.png
```
CozLogo

zamicol commented on coze.go:
```diff
@@ -241,8 +246,13 @@ func (p *Pay) UnmarshalJSON(b []byte) error {
+	}
+
+	// Enforce revoke message max size to prevent DoS attacks.
+	if p2.Rvk > 0 && RVK_MAX_SIZE > 0 && len(b) > RVK_MAX_SIZE {
```
I don't think we need the first greater than here because it's done previously.

zamicol replied:
Naw this is correct.

zamicol commented on coze_test.go:
```diff
@@ -46,7 +46,53 @@ func ExamplePay_embedded() {
+	content := "Coz is a cryptographic JSON messaging specification."
+
+	// Calculate the SHA-256 digest of the content.
+	// ES256's hash algorithm is SHA-256, so dig's algorithm matches alg.
```
TODO ZAMI I want to fix this once it's merged.

zamicol replied:
Done.

zamicol commented on coze_test.go:
```diff
@@ -483,7 +529,40 @@ func Example_iat_rvk_too_big() {
 	// {"rvk":9007199254740991}
 }
+
+// Example demonstrating RVK_MAX_SIZE enforcement for revoke messages.
```
TODO ZAMI look at this once it's merged.

zamicol replied:
The first is for nil.  This is correct.

nrdxp committed:
chore: remove `e` from logo image

zamicol commented on key.go:
```diff
@@ -387,7 +413,7 @@ func (c *Key) Correct() (err error) {
 	}
 
-	// tmb only key.  (Coze assumes `x` is calculable from `d`, so at this point
+	// tmb only key.  (Coze assumes `pub` is calculable from `prv`, so at this point
```
Coz not Coze
